Framework/Algorithms/src/CalcCountRate.cpp

```diff
@@ -0,0 +1,653 @@
+/** Execute the algorithm.
+ */
+void CalcCountRate::exec() {
+
+  DataObjects::EventWorkspace_sptr sourceWS = getProperty("Workspace");
+  API::EventType et = sourceWS->getEventType();
+  if (et == API::EventType::WEIGHTED_NOTIME) {
+    throw std::runtime_error("Event workspace " + sourceWS->name() +
+                             " contains events without necessary frame "
+                             "information. Can not process counting rate");
+  }
+
+  // Identity correct way to treat input logs and general properties of output
+  // log
+  this->setOutLogParameters(sourceWS);
+
+  //-------------------------------------
+  // identify ranges for count rate calculations and initiate source workspace
+  this->setSourceWSandXRanges(sourceWS);
+
+  // check if visualization workspace is necessary and if it is, prepare
+  // visualization workspace to use
+  this->checkAndInitVisWorkspace();
+
+  // create results log and add it to the source workspace
+  std::string logname = getProperty("CountRateLogName");
+  auto newlog = new Kernel::TimeSeriesProperty<double>(logname);
+  sourceWS->mutableRun().addProperty(newlog, true);
+
+  // calculate averages requested and modify results log
+  this->calcRateLog(m_workingWS, newlog);
+
+  // clear up log derivative and existing log pointer (if any)
+  // to avoid incorrect usage
+  // at subsequent calls to the same algorithm.
+  m_tmpLogHolder.release();
+  m_pNormalizationLog = nullptr;
+}
+
+/** Process input workspace to calculate instrument counting rate as function of
+*experiment time
+*@param InputWorkspace :: shared pointer to the input workspace to process
+*@param targLog        :: pointer to time series property containing count rate
+*log.
+*                         Property should exist on input and will be modified
+*with
+*                         counting rate log on output.
+*/
+void CalcCountRate::calcRateLog(
+    DataObjects::EventWorkspace_sptr &InputWorkspace,
+    Kernel::TimeSeriesProperty<double> *const targLog) {
+
+  MantidVec countRate(m_numLogSteps);
+  std::vector<double> countNormalization;
+  if (this->notmalizeCountRate())
+    countNormalization = m_pNormalizationLog->valuesAsVector();
+
+  std::unique_ptr<std::mutex[]> pVisWS_locks;
+  if (this->buildVisWS()) {
+    pVisWS_locks.reset(new std::mutex[m_visWs->getNumberHistograms()]);
+  }
+
+  int64_t nHist = static_cast<int64_t>(InputWorkspace->getNumberHistograms());
+  // Initialize progress reporting.
+  API::Progress prog(this, 0.0, 1.0, nHist);
+
+  double dTRangeMin = static_cast<double>(m_TRangeMin.totalNanoseconds());
+  double dTRangeMax = static_cast<double>(m_TRangeMax.totalNanoseconds());
+  std::vector<MantidVec> Buff;
+  int nThreads;
+
+#pragma omp parallel
+  {
+#pragma omp single
+    {
+      // initialize thread's histogram buffer
+      nThreads = PARALLEL_NUMBER_OF_THREADS;
+      Buff.resize(nThreads);
+      for (int i = 0; i < nThreads; i++) {
+        Buff[i].assign(m_numLogSteps, 0);
```

Comment:
resizing before parallel section will not work, as real number of threads will be known in parallel section only, and before the parallel section its usually one. 

Common memory initialization has to be done on a single thread, otherwise each thread will reallocate memory and the first one will certainly already use memory reallocated by the last one. 

Allocating thread-specific memory indeed can be done by single thread. That's just instinct I follow on allocating memory -- making this change, though as usual with parallel, it makes me think a lot.